README
			=======================
			Not-So-Bad Distribution
			=======================

This is the source release of Not-So-Bad Distribution, an Open WWW-based
Authenticating Software Distribution system.  This is free software, released
under the terms of the GNU General Public License; see the COPYING file for
details.  The home page of NSBD is
    http://www.bell-labs.com/nsbd
Up-to-date NSBD binary packages are available on multiple Unix platforms,
so check the web page first to see if your platform is supported before
compiling it.

NSBD is written primarily in TCL, and you need a TCL8.0 installation or
later to compile it.  You will also need at least "tcl2c" from the TCL
"plus patch", and you will need the entire TCL plus patch if you want to
build a standalone NSBD binary that can run on a system that does not have
TCL installed.  If you also have TK and you have not specified the
--without-tk configure option, TK will be used (without TK you lose the
nsbd -register option); if you want to build a standalone nsbd binary, you
will also need the TK plus patch.  Links to all these packages are on the
NSBD home page.

To compile NSBD on unix, go to the "unix" directory and type "configure"
and "make install".  All the standard GNU autoconf "configure" options
should work, but the only standard option that has been tested is
"--prefix" for the directory prefix to install into.  The NSBD binary is
relocatable since it is completely self-contained, but some scripts that
are useful to package maintainers are installed into a lib directory.

To use NSBD with any kind of security, you will also need PGP or GnuPG.

Most of the documentation is on the home page and in the man page.  The
same information that's available in the man page is available in the
builtin help system in the nsbd program.


ChangeLog
1997/06/19
    First internet release, version 1.0b3.
1997/06/20
    Change nsbdsrc package to have "nsbd/" prefix.
    Change -fetchAll to not apply any local config or registry substitutions.
1997/07/16
    When looking for previously installed '.nsb' files, honor executableTypes
      set on the command line.
1997/08/19
    Support % substitutions in topUrl and multigetUrl.
    Re-organized installations of multiple executableTypes that go into
	separate installTops so that files for all types will be fetched
	together, installed together, and have '.nsb' files installed together.
	No step will occur until the previous step succeeds completely.
1997/08/20
    Added -changedPaths and -batchUpdate options.
    Read at most $CONTENT_LENGTH bytes if the environment variable is set
	whenever "-" is used to read from standard input.
    Fixed bug in -postUrl which caused multi-line POSTs to fail.
1997/08/25
    Added posttonsbd.sh and pushpackage.sh scripts to support direct pushing
	of packages from maintainers to users.
    Added postLength= command line variable for use with -postUrl to avoid
	reading large POST operations into memory, and added corresponding
	-querychannel and -querylength options to http_get.
1997/08/26
    Changed -batchUpdate so it doesn't save all of the updates from standard
	input to a temporary file.
    Improved reports about changes to keys that may be maintainerUpdatableKeys 
	to only show additions or removals of items in a list.  For keys
	that *could* be mantainerUpdatableKeys but aren't, compare against
	values in both the registry database and those saved in any previous
	'.nsb' files.
1997/08/28
    Fixed bug in the building of a '.nsb' from a '.npd' file to make sure
	that files of all executableTypes exist when %E substitutions are
	requested.
    If any paths are missing when building a '.nsb' from a '.npd', now
	prompts the user to ask whether or not to continue without them.
    Print no warnings when creating a '.nsb' file and nsbUrl, topUrl, and
	multigetUrl are all missing, because none of them are needed for
	-batchUpdate.
    Fixed bug with missing nupContents variables when there are no paths.
1997/09/04
    Moved the turning off blocking on the http socket until after a POST
	operation is done, to avoid buffering the data.
    Put in percent-done progress messages on url fetches and -postUrl.
1997/09/05
    Added verbose levels 4 and 6 for no percent-done messages and more
	percent-done messages respectively, and changed the default logLevel
	to 3 and the default verbose level to 5.
    Changed -notifyOnly to -pushOnly in pushpackage.sh.
1997/09/08
    Fixed multiget.sh to work properly on sunos 4.
1997/09/22
    Port to tcl8.0/tk8.0.  Had to put in a workaround in http.tcl to get
	around a bug in tcl8.0 with fconfigure $s -translation {auto binary}.
	There was also a bug in tkConfig.sh's TK_LIB_SPEC for which I did
	not create a workaround but instead I sent in a patch to Sun; this
	only affects some platforms when building with shared libraries.
	Also have to use fcopy instead of unsupported0.
1997/09/23
    Also use fcopy in the support CGI script multiget.tcl.
    Added -changedPaths option to pushpackage.sh.
1997/09/26
    Allow directories to be duplicated after substitutions when installing.
    Fix bug introduced 1997/08/19 which assumed that nonportable files
	in different executable types would always be loaded together.
	That's not necessarily the case if different lists of executable
	types are used on different update operations.
    Changed release number to 1.0, the first non-beta release.
1997/10/03
    Fixed bug in trimOldLog with trying to get the first two columns of
	a log file line using "lrange"; the problem is that the text line
	may not be a well-formed tcl list, for example containing double
	quotes followed by a colon.  Use regexp instead.
1997/10/10
    Fixed bug in urlMultiMdCopyHandler which had a "gets" that ignored the
	fact that the socket is in "-blocking off" mode, which means that
	gets might return -1 if a complete line is not available.
    Also discovered a bug in the plus patches which adds TCL_FILE_EVENTS
	to the list of events that are handled when "update idletasks" is
	called.  This causes errors when loading files with -register
	because file event handlers get nested.   To workaround the bug,
	moved the progress messages to be called just before returning
	from the urlMultiMdCopyHandler so it shouldn't hurt if they
	get nested.
1997/10/13
    Apply local substitutions to validPaths that are explicitly specified
	in a '.npd'/'.nsb' file.
1997/10/14
    Changed proxy specifications to allow a "/" at the end.
    Merged http.tcl with the http1.0 that came with the base tcl8.0 release.
	I noticed that this solved a problem that started when I upgraded
	to tcl8.0, because of a difference in operation between "unsupported0"
	and "fcopy": by default, fcopy will always block until its entire
	-size bytes are received, unless you use the -command option.  As a
	result, operations like nsbd -getUrl would not display results from
	a CGI script as they were happening; the new http.tcl does the
	fcopy with -command.
    Skip all maintainerUpdatableKey processing for keys set on the command
	line; that is, don't update the database (if it is maintainerUpdatable)
	and don't warn about changes (if it is non-maintainerUpdatable).
    In all cases where a '.nsb' file is allowed to be a local filename or
	a URL, also accept a URL specification of file:// because it doesn't
	hurt and may be more intuitive if someone is thinking in terms of
	a URL.
    If topUrl is set on the command line when installing a package and the
	URL is of the form file://, copy files locally instead of fetching
	over the network.  This is useful for re-"distributing" a package
	from one place to another on the same machine.  This is only
	documented under the "-update" option because I can't figure out

generic/sha1.c
/*
SHA-1 in C
By Steve Reid <steve@edmweb.com>
100% Public Domain

Test Vectors (from FIPS PUB 180-1)
"abc"
  A9993E36 4706816A BA3E2571 7850C26C 9CD0D89D
"abcdbcdecdefdefgefghfghighijhijkijkljklmklmnlmnomnopnopq"
  84983E44 1C3BD26E BAAE4AA1 F95129E5 E54670F1
A million repetitions of "a"
  34AA973C D4C4DAA4 F61EEB2B DBAD2731 6534016F
*/

/* If byte order known, #define LITTLE_ENDIAN or BIG_ENDIAN to be faster */
/* #define SHA1HANDSOFF * Copies data before messing with it. */

#include <stdio.h>
#include <string.h>
#include "sha1.h"

#define rol(value, bits) (((value) << (bits)) | ((value) >> (32 - (bits))))

/* blk0() and blk() perform the initial expand. */
/* I got the idea of expanding during the round function from SSLeay */
#ifdef LITTLE_ENDIAN
#define blk0(i) (block->l[i] = (rol(block->l[i],24)&0xFF00FF00) \
    |(rol(block->l[i],8)&0x00FF00FF))
#else
#ifdef BIG_ENDIAN
#define blk0(i) block->l[i]
#else
/* for unknown byte order, to work with either */
/* results in no change on big endian machines */
/* added by Dave Dykstra, 4/16/97 */
#define blk0(i) (block->l[i] = (*(p = (unsigned char *) (&block->l[i])) << 24) \
		+ (*(p+1) << 16) + (*(p+2) << 8) + *(p+3))
#endif
#endif
#define blk(i) (block->l[i&15] = rol(block->l[(i+13)&15]^block->l[(i+8)&15] \
    ^block->l[(i+2)&15]^block->l[i&15],1))

/* (R0+R1), R2, R3, R4 are the different operations used in SHA1 */
#define R0(v,w,x,y,z,i) z+=((w&(x^y))^y)+blk0(i)+0x5A827999+rol(v,5);w=rol(w,30);
#define R1(v,w,x,y,z,i) z+=((w&(x^y))^y)+blk(i)+0x5A827999+rol(v,5);w=rol(w,30);
#define R2(v,w,x,y,z,i) z+=(w^x^y)+blk(i)+0x6ED9EBA1+rol(v,5);w=rol(w,30);
#define R3(v,w,x,y,z,i) z+=(((w|x)&y)|(w&x))+blk(i)+0x8F1BBCDC+rol(v,5);w=rol(w,30);
#define R4(v,w,x,y,z,i) z+=(w^x^y)+blk(i)+0xCA62C1D6+rol(v,5);w=rol(w,30);


/* Hash a single 512-bit block. This is the core of the algorithm. */

#ifdef _USING_PROTOTYPES_
void SHA1Transform(unsigned long state[5], unsigned char buffer[64])
#else
void SHA1Transform(state, buffer)
    unsigned long state[5];
    unsigned char buffer[64];
#endif
{
#if (!defined(BIG_ENDIAN) && !defined(LITTLE_ENDIAN))
unsigned char *p;
#endif
unsigned long a, b, c, d, e;
typedef union {
    unsigned char c[64];
    unsigned long l[16];
} CHAR64LONG16;
CHAR64LONG16* block;
#ifdef SHA1HANDSOFF
static unsigned char workspace[64];
    block = (CHAR64LONG16*)workspace;
    memcpy(block, buffer, 64);
#else
    block = (CHAR64LONG16*)buffer;
#endif
    /* Copy context->state[] to working vars */
    a = state[0];
    b = state[1];
    c = state[2];
    d = state[3];
    e = state[4];
    /* 4 rounds of 20 operations each. Loop unrolled. */
    R0(a,b,c,d,e, 0); R0(e,a,b,c,d, 1); R0(d,e,a,b,c, 2); R0(c,d,e,a,b, 3);
    R0(b,c,d,e,a, 4); R0(a,b,c,d,e, 5); R0(e,a,b,c,d, 6); R0(d,e,a,b,c, 7);
    R0(c,d,e,a,b, 8); R0(b,c,d,e,a, 9); R0(a,b,c,d,e,10); R0(e,a,b,c,d,11);
    R0(d,e,a,b,c,12); R0(c,d,e,a,b,13); R0(b,c,d,e,a,14); R0(a,b,c,d,e,15);
    R1(e,a,b,c,d,16); R1(d,e,a,b,c,17); R1(c,d,e,a,b,18); R1(b,c,d,e,a,19);
    R2(a,b,c,d,e,20); R2(e,a,b,c,d,21); R2(d,e,a,b,c,22); R2(c,d,e,a,b,23);
    R2(b,c,d,e,a,24); R2(a,b,c,d,e,25); R2(e,a,b,c,d,26); R2(d,e,a,b,c,27);
    R2(c,d,e,a,b,28); R2(b,c,d,e,a,29); R2(a,b,c,d,e,30); R2(e,a,b,c,d,31);
    R2(d,e,a,b,c,32); R2(c,d,e,a,b,33); R2(b,c,d,e,a,34); R2(a,b,c,d,e,35);
    R2(e,a,b,c,d,36); R2(d,e,a,b,c,37); R2(c,d,e,a,b,38); R2(b,c,d,e,a,39);
    R3(a,b,c,d,e,40); R3(e,a,b,c,d,41); R3(d,e,a,b,c,42); R3(c,d,e,a,b,43);
    R3(b,c,d,e,a,44); R3(a,b,c,d,e,45); R3(e,a,b,c,d,46); R3(d,e,a,b,c,47);
    R3(c,d,e,a,b,48); R3(b,c,d,e,a,49); R3(a,b,c,d,e,50); R3(e,a,b,c,d,51);
    R3(d,e,a,b,c,52); R3(c,d,e,a,b,53); R3(b,c,d,e,a,54); R3(a,b,c,d,e,55);
    R3(e,a,b,c,d,56); R3(d,e,a,b,c,57); R3(c,d,e,a,b,58); R3(b,c,d,e,a,59);
    R4(a,b,c,d,e,60); R4(e,a,b,c,d,61); R4(d,e,a,b,c,62); R4(c,d,e,a,b,63);
    R4(b,c,d,e,a,64); R4(a,b,c,d,e,65); R4(e,a,b,c,d,66); R4(d,e,a,b,c,67);
    R4(c,d,e,a,b,68); R4(b,c,d,e,a,69); R4(a,b,c,d,e,70); R4(e,a,b,c,d,71);
    R4(d,e,a,b,c,72); R4(c,d,e,a,b,73); R4(b,c,d,e,a,74); R4(a,b,c,d,e,75);
    R4(e,a,b,c,d,76); R4(d,e,a,b,c,77); R4(c,d,e,a,b,78); R4(b,c,d,e,a,79);
    /* Add the working vars back into context.state[] */
    state[0] += a;
    state[1] += b;
    state[2] += c;
    state[3] += d;
    state[4] += e;
    /* Wipe variables */
    a = b = c = d = e = 0;
}


/* SHA1Init - Initialize new context */

#ifdef _USING_PROTOTYPES_
void SHA1Init(SHA1_CTX* context)
#else
void SHA1Init(context)
    SHA1_CTX* context;
#endif
{
    /* SHA1 initialization constants */
    context->state[0] = 0x67452301;
    context->state[1] = 0xEFCDAB89;
    context->state[2] = 0x98BADCFE;
    context->state[3] = 0x10325476;
    context->state[4] = 0xC3D2E1F0;
    context->count[0] = context->count[1] = 0;
}


/* Run your data through this. */

#ifdef _USING_PROTOTYPES_
void SHA1Update(SHA1_CTX* context, unsigned char* data, unsigned int len)
#else
void SHA1Update(context, data, len)
    SHA1_CTX* context;
    unsigned char* data;
    unsigned int len;
#endif
{
unsigned int i, j;

    j = (context->count[0] >> 3) & 63;
    if ((context->count[0] += len << 3) < (len << 3)) context->count[1]++;
    context->count[1] += (len >> 29);
    if ((j + len) > 63) {
        memcpy(&context->buffer[j], data, (i = 64-j));
        SHA1Transform(context->state, context->buffer);
        for ( ; i + 63 < len; i += 64) {
#ifdef CANWRITEDATA
            SHA1Transform(context->state, &data[i]);
#else
	    /* else case added by Dave Dykstra 4/22/97 */
	    memcpy(&context->buffer[0], &data[i], 64);
            SHA1Transform(context->state, context->buffer);
#endif
        }
        j = 0;
    }
    else i = 0;
    memcpy(&context->buffer[j], &data[i], len - i);
}


/* Add padding and return the message digest. */

#ifdef _USING_PROTOTYPES_
void SHA1Final(unsigned char digest[20], SHA1_CTX* context)
#else
void SHA1Final(digest, context)
    unsigned char digest[20];
    SHA1_CTX* context;
#endif
{
unsigned long i, j;
unsigned char finalcount[8];

    for (i = 0; i < 8; i++) {
        finalcount[i] = (unsigned char)((context->count[(i >= 4 ? 0 : 1)]
         >> ((3-(i & 3)) * 8) ) & 255);  /* Endian independent */
    }
    SHA1Update(context, (unsigned char *)"\200", 1);
    while ((context->count[0] & 504) != 448) {
        SHA1Update(context, (unsigned char *)"\0", 1);
    }
    SHA1Update(context, finalcount, 8);  /* Should cause a SHA1Transform() */
    for (i = 0; i < 20; i++) {
        digest[i] = (unsigned char)
         ((context->state[i>>2] >> ((3-(i & 3)) * 8) ) & 255);
    }
    /* Wipe variables */
    i = j = 0;
    memset(context->buffer, 0, 64);
    memset(context->state, 0, 20);
    memset(context->count, 0, 8);
    memset(&finalcount, 0, 8);
#ifdef SHA1HANDSOFF  /* make SHA1Transform overwrite it's own static vars */
    SHA1Transform(context->state, context->buffer);
#endif
}


#ifdef DO_MAIN
/*************************************************************/


#ifdef _USING_PROTOTYPES_
int main(int argc, char** argv)
#else
int main(argc, ar** argv)
    int argc;
    char** argv;
#endif
{
int i, j;
SHA1_CTX context;
unsigned char digest[20], buffer[16384];
FILE* file;

    if (argc > 2) {
        puts("Public domain SHA-1 implementation - by Steve Reid <steve@edmweb.com>");
        puts("Produces the SHA-1 hash of a file, or stdin if no file is specified.");
        exit(0);
    }
    if (argc < 2) {
        file = stdin;
    }
    else {
        if (!(file = fopen(argv[1], "rb"))) {
            fputs("Unable to open file.", stderr);
            exit(-1);
        }
    } 
    SHA1Init(&context);
    while (!feof(file)) {  /* note: what if ferror(file) */
        i = fread(buffer, 1, 16384, file);
        SHA1Update(&context, buffer, i);
    }
    SHA1Final(digest, &context);
    fclose(file);
    for (i = 0; i < 5; i++) {
        for (j = 0; j < 4; j++) {
            printf("%02X", digest[i*4+j]);
        }
        putchar(' ');
    }
    putchar('\n');
    exit(0);
}
#endif
